existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+For example, it is very likely from the function definition that `m` is an instance of `Monad` typeclass.
+In order to add this constraint we need to place it after all variables are declared:
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+``` 
+You might want to ask, what is so special about this syntax?
+Nothing changed since we added the quantifier.
+Well, actually there are few things to notice here.
+
+### Examples
+#### Order of type variables
+First of all, now you can change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (Yeah, there are real life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.
+Without forall you will write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+``` 
+
+Quite long right?
+This is because ghc will maintain the same order, as these variables first appear in function definition.
+However, using explicit `forall.` this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.
+
+#### Additional type variables
+Second, you now can define functions with more type variables than you use on the right side of declaration.
+This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:
+
+``` Haskell
+someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
+```
+
+Here we don't have `env` used in the type definition, but despite that, we are still able to add `MonadReader` constraint.
+
+## Existential quantification
+Beside universally quantified variables, Haskell also supports existential types.
+In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
+That might be a little confusing.
+The following construction is called existential type in Haskell - when you use a forall quantifier in the right-hand side of function or data type declaration.
+But what is it used for? Let's find out!
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
+```
+Few things to mention.
+Here `a` and `c` are universally quantified and `b` is existentially quantified variables.
+The same variable can't be quantified more than once.
+Also, you can see that constraints for `b` are added in the middle of the function type declaration.
+This example is pretty useless.
```

Comment:
Are you saying that this example in particular is useless or that existential type variables in function declarations (rather than in data declarations) are useless?

Either way, if this example is useless, we should come up with something more realistic that better demonstrates its usefulness.

Off the top of my head:

```hs
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do

  username <- getUsername
  log username
  
  userCount <- getUserCount
  log userCount

getUsername :: IO String
getUsername = undefined

getUserCount :: IO Int
getUserCount = undefined
```

```hs
main :: IO ()
main = do
  -- log to the console
  func print

  -- log to a logfile
  func (appendFile "logfile.log" . show)
```